components/brave_wallet/browser/simulation_service.cc

```diff
@@ -0,0 +1,326 @@
+  GURL url(spec);
+  return url;
+}
+
+void SimulationService::HasTransactionScanSupport(
+    const std::string& chain_id,
+    mojom::CoinType coin,
+    HasTransactionScanSupportCallback callback) {
+  bool result = HasTransactionScanSupportInternal(chain_id, coin);
+  std::move(callback).Run(result);
+}
+
+void SimulationService::HasMessageScanSupport(
+    const std::string& chain_id,
+    mojom::CoinType coin,
+    HasMessageScanSupportCallback callback) {
+  bool result = HasMessageScanSupportInternal(chain_id, coin);
+  std::move(callback).Run(result);
+}
+
+void SimulationService::ScanSolanaTransaction(
+    mojom::SolanaTransactionRequestUnionPtr request,
+    const std::string& language,
+    ScanSolanaTransactionCallback callback) {
+  const auto& chain_id =
+      json_rpc_service_->GetChainIdSync(mojom::CoinType::SOL, absl::nullopt);
+
+  if (!HasTransactionScanSupportInternal(chain_id, mojom::CoinType::SOL)) {
+    std::move(callback).Run(
+        nullptr, "",
+        l10n_util::GetStringUTF8(IDS_BRAVE_WALLET_UNSUPPORTED_NETWORK));
+    return;
+  }
+
+  const auto& encoded_params =
+      solana::EncodeScanTransactionParams(std::move(request));
+  if (!encoded_params) {
+    std::move(callback).Run(
+        nullptr, "", l10n_util::GetStringUTF8(IDS_WALLET_INTERNAL_ERROR));
+    return;
+  }
+
+  auto internal_callback =
+      base::BindOnce(&SimulationService::OnScanSolanaTransaction,
+                     weak_ptr_factory_.GetWeakPtr(), std::move(callback));
+
+  auto conversion_callback = base::BindOnce(&ConvertAllNumbersToString);
+
+  api_request_helper_.Request(
+      "POST", GetScanTransactionURL(chain_id, mojom::CoinType::SOL),
+      *encoded_params, "application/json", std::move(internal_callback),
+      GetHeaders(), {.auto_retry_on_network_change = true},
+      std::move(conversion_callback));
+}
+
+void SimulationService::OnScanSolanaTransaction(
+    ScanSolanaTransactionCallback callback,
+    APIRequestResult api_request_result) {
+  if (!api_request_result.Is2XXResponseCode()) {
+    if (auto error_response =
+            ParseSimulationErrorResponse(api_request_result.value_body())) {
+      std::move(callback).Run(nullptr, std::move(*error_response), "");
+    } else {
+      std::move(callback).Run(
+          nullptr, "", l10n_util::GetStringUTF8(IDS_WALLET_PARSING_ERROR));
+    }
+
+    return;
+  }
+
+  if (auto simulation_response =
+          solana::ParseSimulationResponse(api_request_result.value_body())) {
+    std::move(callback).Run(std::move(simulation_response), "", "");
+  } else {
+    std::move(callback).Run(nullptr, "",
+                            l10n_util::GetStringUTF8(IDS_WALLET_PARSING_ERROR));
+  }
+}
+
+void SimulationService::ScanEVMTransaction(
+    mojom::TransactionInfoPtr tx_info,
+    const std::string& language,
+    ScanEVMTransactionCallback callback) {
+  const auto& chain_id =
+      json_rpc_service_->GetChainIdSync(mojom::CoinType::ETH, absl::nullopt);
```

Comment:
same as https://github.com/brave/brave-core/pull/18302/files#r1183100154, we need to obtain `chainId` from `TransactionInfo`

Reply:
Fixed in 012cb657ca1aa9a22c14fe0497d8e514f5a8e3eb.